### DISCUSSION TITLE
Replace playlist not working in 20210402 update

### DISCUSSION THREAD

**Discussion by JoeS209** [rern/rAudio#286] [Fixed Issues]
My general use pattern is to browse to a folder in the library view, then Replace/Play that album/folder.  Starting with the 20210402 update, the playlist is not being replaced, and playing starts at the old playlist.  Note that "Consume Mode" is Off in my playlist.

**Comment by JoeS209** [comment]
Even worse, if I clear the playlist first and then try to play a new folder, the contents of the playlist that was deleted reappear!

**Comment by JoeS209** [comment]
Ah, what I think is really happening here, is that choosing to "Play/Replace" a folder is actually recursively adding everything from the *parent* folder and below.

**Comment by EarsofTin** [comment]
I can confirm the same issue, with or without consume mode.  The current queue is augmented or replaced by all the tracks from all subfolders in a folder when choosing a single subfolder.

**Comment by EarsofTin** [comment]
Problem persists with fresh image updated to 20210403. This is a difficult issue for me in particular as I have large collection with lots of compilations and live recordings.  Hence album and artist tags are shabby.  I play my collection from file directories.

**Reply by JoeS209** [reply]
Problem seems to be fixed for me after 20210403 update.

**Reply by EarsofTin** [reply]
Working for me now after repeatedly doing the same action and hoping for a different result (i.e. insanity).  I was already on 20210403 with the problem. Maybe "update database" cleared some debris after remounting my NAS.

**Reply by EarsofTin** [reply]
Very strange!  Problem occurs on different control devices and has to be fixed on each. It resolved on my tablet.  Then I used my phone and re-experienced the problem.  This is what fixes it for me.  Open any folder to the individual track listing.  Add any track to the queue.  The first time nothing will be added to the queue.  Repeat.  Then it works.  After that you can add an entire folder (and just that folder) to the queue as normal.  Repeat for each browser you control the player from.

**Reply by JoeS209** [reply]
Ah, maybe clear your browser caches.

**Reply by EarsofTin** [reply]
Unfortunately, clearing the cache seems to reinstate the problem.

**Comment by JoeS209** [comment]
I can confirm that the problem just happened again for me.  I am running 20210403.

**Comment by EarsofTin** [comment]
Problem continues using update 20210403.  Tried direct control instead of headless and different browsers but no success.  Temporary workaround is to add a single track then replace with the entire directory but that fix is lost when the browser is closed.  I reverted to the 20210330 image which does not exhibit the problem and I will avoid further updates.  Looking for the funny in this situation, the description of the 20210403 update claims to fix the problem I wasn't having before the update.  Still the best player software available.  Thanks to the developer for a great product.

**Comment by kosto-stavr** [comment]
I can confirm too that the problem continues using update 20210403.

**Comment by rern** [comment]
Try Add/Replace with SSH command:

1. Add code
```sh
str=$( cat <<EOF
\\\tinfo( { message : 'cmd.sh "'+ command.toString().replace( /,/g, '<br>' ) +'"', msgalign : 'left' } ); return
EOF
)
sed -i "/function addReplace/ a$str" /srv/http/assets/js/context.js
```

2. Add/Replace Playlist
- Refresh rAudio browser
- From Library page, Add/Replace a directory/folder
- Copy command from popup
- Open SSH terminal > Paste the command
- See Playlist page

3. Once done, remove the added code
```sh
sed -i '/command.toString/ d' /srv/http/assets/js/context.js
```

4. Post the command and errors.

_(Edited: Fix Add/Remove code to include newlines)_

**Reply by EarsofTin** [reply]
Not sure I'm following your intent but this is what happened:

 cmd.sh "pllsNAS/quarantineaddplay0.2"

After entering command via SSH the playlist page is empty, no error messages displayed, problem persists.  "Quarantine" is the name of a small directory that I use for testing setups and it contains several subfolders ("albums").

**Reply by rern** [reply]
Just found out that `window.prompt` on Firefox drops newline characters `\n` form the command.

Try again with revised `Add code` which uses rAudio popup instead.

By the way, the ditectory name: `quarantine` or `Quarantine`?

**Reply by EarsofTin** [reply]
cmd.sh "plls
NAS/quarantine"

now appears in an information box instead of a command box.  The directory is named "quarantine".  At the moment I'm using Firefox but I also use Chrome and Brave.  No explicit errors are displayed.

**Reply by rern** [reply]
Behind the scene, the command does:
```sh
mpc ls "NAS/quarantine" | mpc add
```
So try `mpc ls "NAS/quarantine"`. It should list all directories/files.

**Reply by EarsofTin** [reply]
The command lists the subdirectories, no files

rAudio:~  #  mpc ls "NAS/quarantine"
NAS/quarantine/Billy Walton Band - Billy Walton Band
NAS/quarantine/Billy Walton Band - BWB studio sampler
NAS/quarantine/Billy Walton Band - Crank It Up
NAS/quarantine/Billy Walton Band - Live at the Stone Pony
NAS/quarantine/Billy Walton Band - Neon City
NAS/quarantine/Dark Hour
NAS/quarantine/Live In the U.K
NAS/quarantine/Soul of a Man
NAS/quarantine/Wish for What You Want

edited as I am trying different levels of file tagging

**Reply by rern** [reply]
Try `UPDATE` branch:
- Addons page > rAudio 1 - long-press `Update` button
- OK for `UPDATE` branch

**Reply by EarsofTin** [reply]
No luck, the update recreated the problem.  I had reverted to the 20210330 image which doesn't exhibit the problem.  Problem first observed with 20210403 update which I guess includes 20210402 update.  Is there a way to add one update at a time or avoid single updates?  I spent some time yesterday looking at the code for each update but I'm way over my head.

Temporary workaround of adding single track then replacing with full directory usually works but sometimes add/replace gives null result.  Temporary fix until browser closed.

**Reply by rern** [reply]
- Try the UPDATE branch again.
- If still no, try copying an albums to /mnt/MPD/SD.
- Update Library then try add/replace from SD.

**Reply by kaufmach** [reply]
I experience the same issue. Did also a re-install yesterday. Behavior remains the same. But workaround suggested by EarsofTin works for now. Tried also, of course, update whole library. I did not install anything else. Only using USB music library and webradio with Ethernet. Just wanted to let you know ...

Also, I do like rAudio1 (tried several in the past, Rune versions before rAudio1, Moodeaudio, Volumio, and other approaches): Thanks so much for your work, awesome!

System:
rAudio 1 • 20210416
5.10.27-3-ARCH armv7l
Raspberry Pi 4 Model B Rev 1.2
/dev/mmcblk0p2 2.4GB/30GB
/mnt/MPD/USB/platte • /dev/sda1 1.4TB/1.9TB

**Reply by EarsofTin** [reply]
Fresh install of 20210330.  Added folders(albums) to the SD card and attached USB thumbdrive.  Then rebuilt library.  The add/replace function worked for all three sources - SD/USB/NAS.

Then I ran 20210416 update.  Add/Replace now malfunctions for all three sources i.e. adding a single folder(album) adds all folders in that directory.

**Comment by EarsofTin** [comment]
This problem now solved.

I'm running a fresh 64 bit image with updates (rAudio 1 • 20210502 5.10.27-3-ARCH aarch64).  I don't know if it's 64bit or updates but I'm happy.

Thanks Rern!
